cts/scheduler/xml/bundle-promoted-location-5.xml

```diff
@@ -0,0 +1,225 @@
+<cib crm_feature_set="3.17.4" validate-with="pacemaker-3.9" epoch="134" num_updates="40" admin_epoch="0" cib-last-written="Thu Jun 22 17:00:31 2023" update-origin="node1" update-client="cibadmin" update-user="root" have-quorum="1" dc-uuid="1">
+  <configuration>
+    <!-- The essential elements of this test are:
+         * A promotable bundle (base-bundle) is promoted on node3 and unpromoted
+           on node1 and node2
+         * There is a positive location constraint (score=5000) for the bundle's
+           promoted role, preferring node2
+         * There are no other constraints
+         * There is no stickiness
+
+         In this situation, the bundle should demote on node3 and promote on
+         some other node.
+
+         This test is incorrect. The bundle remains promoted on node3 and
+         unpromoted on other nodes.
```

Comment:
Do you think this one is still incorrect? Positive location on `node3` for unpromoted role, bundle should demote on `node3` and promote somewhere else.

Reply:
Ah, it's fine. I remembered where I saw it:

https://github.com/ClusterLabs/pacemaker/blob/main/lib/pacemaker/pcmk_sched_location.c#L121-L127

Reply:
This case is arguably still broken then, depending on how we want it to behave.

The code you linked remaps unpromoted to unknown. Which means if we have a positive preference for unpromoted on a given node, that node still won't prefer to be unpromoted (vs. promoted). That's iffy, but I would think all else being equal, positive preference for unpromoted should mean we prefer to keep that node unpromoted. Whereas banning the unpromoted role is a complete ban, like you said in another comment.

It also looks like we never check for role == RSC_ROLE_STOPPED. So even though we don't document that role as allowed, if someone specifies it, things might not work right. Will add a to-do note, to do a test later.

Reply:
I think it makes sense. If someone wants a node to not be promoted, they should use a rule for the promoted role with a negative score; if they want a node not to be used, they should use a regular constraint (implicitly for started role) with a negative score.

The way I view it is that a location preference for the unpromoted role doesn't imply that the instance needs to *stay* unpromoted.

(We should probably disallow location constraints for the Stopped role. Mapping it to a negative location could make sense, but what would a location for Stopped with a negative score logically mean?)

Reply:
Using a Promoted-role constraint with a negative score makes sense. I think the fact that we support specifying the Unpromoted role intuitively suggests the "all else being equal, prefer to run an Unpromoted instance here" reading though.

Users may also be reluctant to specify a negative score for the Promoted role because net-negative often means "cannot" rather than "prefer not," with current Pacemaker.

It's tricky to specify and it may be tricky to implement in the "prefer Unpromoted" way. I'll alter the comment in this test, since it's not clearly incorrect.

---

For Stopped role: Yeah... negative location would make sense. And though it would be odd, Stopped with negative score would mean "prefer not to be stopped on this node." Which I guess could come into play because any of the other three roles could satisfy it?

It would be odd to set a positive Stopped constraint and even more so to set a negative Stopped constraint, I agree. The meaning is similar for Stopped constraints with colocations, which we added support for years ago and never documented in Pacemaker Explained. I guess that would be like an anticolocation.

Reply:
yeah Stopped is a whole other can of worms. probably worth ignoring for now.

I could see location for unpromoted meaning prefer to promote elsewhere, but that just seems like a headache compared to just treating it like started. I'd rather look at it as passing through the unpromoted role being sufficient to satisfy the constraint. We could convert it internally to a negative colocation for the promoted role, but again I think it isn't worth the hassle.